### PR TITLE
(WIP) jwt conform to standard +  JWT authenticator improvements

### PR DESCRIPTION
**Pull request by jmcardon**
- [x] Turn `jwtId` into an `Option`. Conforms to the standard without needing to break the jwt authenticator.
- [x] `SecureRandomId` now has `Interactive` and `Strong` default vals. Each allocate a `SecureRandom` instance so they are lazy vals.
- [x] JWT authenticator to allow completely stateless objs.
- [x] Possibly: JWT authenticator creation to become more generic.

### CODE REVIEW

File: tsec-http4s/src/main/scala/tsec/authentication/SignedCookieAuthenticator.scala [hunk 1]
```diff
@@ -15,10 +15,118 @@ import tsec.mac.jca._
...
+
+  val expiry: FiniteDuration = settings.expiryDuration
+
+  val maxIdle: Option[FiniteDuration] = settings.maxIdle
+
+  /** Generate a nonce by concatenating the message to be sent with the current instant and hashing their result
+    * Possibly this should be a variable argument, but for now SHA1 is enough, since the chance for collision is
+    * abysmally low.
+    */
+  private def generateNonce(message: String) =
```
Comment (rsoeldner) [inline tsec-http4s/src/main/scala/tsec/authentication/SignedCookieAuthenticator.scala]: should return `F[]`

File: tsec-http4s/src/main/scala/tsec/authentication/SignedCookieAuthenticator.scala [hunk 2]
```diff
@@ -15,10 +15,118 @@ import tsec.mac.jca._
 
 import scala.concurrent.duration.FiniteDuration
 
-abstract class SignedCookieAuthenticator[F[_]: Sync, I, V, Alg] private[tsec] (
-    val expiry: FiniteDuration,
-    val maxIdle: Option[FiniteDuration]
-) extends AuthenticatorService[F, I, V, AuthenticatedCookie[Alg, I]]
+abstract class SignedCookieAuthenticator[F[_], I, V, Alg] private[tsec] (
```
Comment (rsoeldner) [inline tsec-http4s/src/main/scala/tsec/authentication/SignedCookieAuthenticator.scala]: `Alg = A`

File: tsec-http4s/src/test/scala/tsec/authentication/JWTAuthenticatorSpec.scala [hunk 3]
```diff
@@ -15,80 +14,95 @@ import scala.concurrent.duration._
...
     val macKey     = macKeyGen.generateKey
-    val authenticator = JWTAuthenticator.withBackingStoreArbitrary[IO, Int, DummyUser, A](
-      settings,
-      store,
-      dummyStore,
-      macKey
-    )
+    val authenticator: JWTAuthenticator[IO, Int, DummyUser, A] =
+      tf(store, dummyStore, macKey)
+//      tf(settings.expiryDuration, settings.maxIdle, store, dummyStore, macKey)
```
Comment (rsoeldner) [inline tsec-http4s/src/test/scala/tsec/authentication/JWTAuthenticatorSpec.scala]: this line should stay ?

File: tsec-http4s/src/test/scala/tsec/authentication/JWTAuthenticatorSpec.scala [hunk 4]
```diff
@@ -148,100 +158,45 @@ class JWTAuthenticatorSpec extends RequestAuthenticatorSpec {
...
       implicit cv: JWSMacCV[IO, A],
-      enc: AESCTR[E],
-      idKeyGen: IdKeyGen[E, SecretKey],
       macKeyGen: IdKeyGen[A, MacSigningKey]
-  ): AuthSpecTester[AugmentedJWT[A, Int]] = {
-    implicit val instance = enc.genEncryptor[IO].unsafeRunSync()
-    implicit val strategy = enc.defaultIvStrategy[IO]
+  ): StatelessSpecTester[AugmentedJWT[A, DummyUser]] = {
+    val macKey = macKeyGen.generateKey
+    //      settings.expiryDuration,
```
Comment (rsoeldner) [inline tsec-http4s/src/test/scala/tsec/authentication/JWTAuthenticatorSpec.scala]: same

File: tsec-http4s/src/main/scala/tsec/authentication/AuthToken.scala [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  */
+trait AuthToken[A] {
+  def expiry(a: A): Instant
+  def lastTouched(a: A): Option[Instant]
+
+  def isExpired(a: A, now: Instant): Boolean = expiry(a).isBefore(now)
+  def isTimedOut(a: A, now: Instant, timeOut: FiniteDuration): Boolean =
+    lastTouched(a).exists(
+      _.plusSeconds(timeOut.toSeconds)
+        .isBefore(now)
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply tsec-http4s/src/main/scala/tsec/authentication/AuthToken.scala]: Scalafmt lol

[the remark this answers is not in the document]
Comment (jmcardon) [reply tsec-http4s/src/main/scala/tsec/authentication/AuthToken.scala]: Scalafmt lol

File: tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala [hunk 6]
```diff
@@ -8,7 +8,7 @@ import org.log4s._
 import tsec.authorization._
 
 sealed abstract class SecuredRequestHandler[F[_], Identity, User, Auth](
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala]: Sure. This is left over from before, but yes

File: tsec-http4s/src/main/scala/tsec/authentication/SignedCookieAuthenticator.scala (hunk 1 above)
Comment (jmcardon) [reply tsec-http4s/src/main/scala/tsec/authentication/SignedCookieAuthenticator.scala]: Actually no. Hashing on the JCA at least is pure. This doesn't mutate anything

File: tsec-http4s/src/test/scala/tsec/authentication/JWTAuthenticatorSpec.scala (hunk 3 above)
Comment (jmcardon) [reply tsec-http4s/src/test/scala/tsec/authentication/JWTAuthenticatorSpec.scala]: Leftover from heavy refactoring, sorri

File: tsec-http4s/src/test/scala/tsec/authentication/JWTAuthenticatorSpec.scala (hunk 4 above)
Comment (jmcardon) [reply tsec-http4s/src/test/scala/tsec/authentication/JWTAuthenticatorSpec.scala]: Leftover. Refactored hard

File: tsec-http4s/src/main/scala/tsec/authentication/SignedCookieAuthenticator.scala (hunk 1 above)
Comment (jmcardon) [reply tsec-http4s/src/main/scala/tsec/authentication/SignedCookieAuthenticator.scala]: I take that back. You are absolutely right about this. I didn't see the Instant.now call, i wrote that a while ago. My bad. Will fix

File: tsec-http4s/src/main/scala/tsec/authentication/JWTAuthenticator.scala [hunk 7]
```diff
@@ -336,477 +204,361 @@ object JWTAuthenticator {
...
     }
 
-  /** Create a JWT with an encrypted user Id in the `subject` claim,
-    * transported in an arbitrary header
-    *
-    */
-  def statelessEncryptedArbitrary[F[_], I: Decoder: Encoder, V, A: JWTMacAlgo: JCAMacTag, E](
-      settings: TSecJWTSettings,
-      identityStore: IdentityStore[F, I, V],
+  private[tsec] def embedded[F[_], V: Decoder: ObjectEncoder, A: JWTMacAlgo: JCAMacTag](
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply tsec-http4s/src/main/scala/tsec/authentication/JWTAuthenticator.scala]: I removed the stateless encrypted ones. Im not convinced they are misuse proof enough for me to promote using them

File: tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmcardon) [reply tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala]: Actually, I take that back. I feel like this signature might be read more often to understand what's happening. In that scenario, `Identity`, `User`, `Auth` is verbose and I like it.

**Pull request merged**